ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PrivacyBlockProcessor.java

```diff
@@ -95,31 +118,97 @@ public Result processBlock(
                             .getPrivateTransaction()
                             .getPrivacyGroupId()
                             .get());
-                if (!preProcessPrivacyGroupHeadBlockMap.containsKey(privacyGroupId)) {
+
+                final List<PrivateTransactionWithMetadata> actualList =
+                    createActualList(
+                        blockHeader, privateTransactionWithMetadataList, privacyGroupId);
+
+                if (actualList.size() > 0) {
+                  LOG.info(
+                      "Rehydrating privacy group {}, number of transactions to be rehydrated is {} out of a total number of {} transactions.",
+                      privacyGroupId.toString(),
+                      actualList.size(),
+                      privateTransactionWithMetadataList.size());
                   final PrivateStateRehydration privateStateRehydration =
                       new PrivateStateRehydration(
                           privateStateStorage,
                           blockchain,
                           protocolSchedule,
                           publicWorldStateArchive,
                           privateWorldStateArchive);
-                  privateStateRehydration.rehydrate(privateTransactionWithMetadataList);
+                  privateStateRehydration.rehydrate(actualList);
                   privateStateStorage.updater().putAddDataKey(privacyGroupId, addKey).commit();
                 }
               } catch (final EnclaveClientException e) {
-                // we were not being added
+                // we were not being added because we have not found the add blob
               }
             });
-    final PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
-        new PrivacyGroupHeadBlockMap(
+  }
+
+  private List<PrivateTransactionWithMetadata> createActualList(
```

Comment:
I feel like if we move this logic into its own class (e.g. `PrivateTransactionRehydrationSelector` (or something with a better name) we could have multiple unit tests to guarantee the different possible logical paths.

And the unit tests could help one understand the different scenarios.

Maybe we don't do it now, but it would be a good thing.

Reply:
I think there is some rehydration code in the DefaultPrivacyController as well! 
I think we could have a class that contains a lot of that rehydration code, like creation of the blob, creating this list ...
I think we should do that when we are doing the refactor of EeaSendRawTransaction ...